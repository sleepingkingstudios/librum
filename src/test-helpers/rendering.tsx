import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
