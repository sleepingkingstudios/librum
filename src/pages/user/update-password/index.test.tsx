import * as React from 'react';

import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { UserUpdatePassword } from './index';
import { render } from '@test-helpers/rendering';

describe('<UserUpdatePassword />', () => {
  it('should show the link', () => {
    const { getByText } = render(
      <UserUpdatePassword />,
      { store: true },
    );
    const link = getByText('Change Password');

    expect(link).toBeVisible();
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <UserUpdatePassword />,
      { store: true },
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('when the user clicks the Change Password link', () => {
    it('should display the change password form', async () => {
      const { getByRole, getByText } = render(
        <UserUpdatePassword />,
        { store: true },
      );
      const link = getByText('Change Password');

      await userEvent.click(link);

      const button = getByRole('button', { name: 'Update Password' });

      expect(button).toBeVisible();
    });

    it('should match the snapshot', async () => {
      const { asFragment, getByText } = render(
        <UserUpdatePassword />,
        { store: true },
      );
      const link = getByText('Change Password');

      await userEvent.click(link);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
