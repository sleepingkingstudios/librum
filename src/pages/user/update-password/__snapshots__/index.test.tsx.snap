// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserUpdatePassword /> should match the snapshot 1`] = `
<DocumentFragment>
  <a>
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-user-lock pr-1"
      data-icon="user-lock"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 640 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H392.6c-5.4-9.4-8.6-20.3-8.6-32V352c0-2.1 .1-4.2 .3-6.3c-31-26-71-41.7-114.6-41.7H178.3zM528 240c17.7 0 32 14.3 32 32v48H496V272c0-17.7 14.3-32 32-32zm-80 32v48c-17.7 0-32 14.3-32 32V480c0 17.7 14.3 32 32 32H608c17.7 0 32-14.3 32-32V352c0-17.7-14.3-32-32-32V272c0-44.2-35.8-80-80-80s-80 35.8-80 80z"
        fill="currentColor"
      />
    </svg>
    Change Password
  </a>
</DocumentFragment>
`;

exports[`<UserUpdatePassword /> when the user clicks the Change Password link should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="w-full lg:w-1/2 relative"
  >
    <div
      class=""
    >
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="old_password"
        >
          Old Password
        </label>
        <input
          class="form-input"
          id="old_password"
          name="oldPassword"
          type="password"
          value=""
        />
      </div>
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="new_password"
        >
          New Password
        </label>
        <input
          class="form-input"
          id="new_password"
          name="newPassword"
          type="password"
          value=""
        />
      </div>
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="confirm_password"
        >
          Confirm Password
        </label>
        <input
          class="form-input"
          id="confirm_password"
          name="confirmPassword"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="gap-3 sm:grid sm:grid-cols-1 lg:grid-cols-2"
    >
      <button
        class="form-button"
        type="button"
      >
        Cancel
      </button>
      <button
        class="form-button button-info"
        type="submit"
      >
        Update Password
      </button>
    </div>
  </form>
</DocumentFragment>
`;
